### ISSUE TITLE
Deserialization of Picture

### ISSUE THREAD

**Issue by yskwong** [ahome-it/lienzo-core#56]
I have run into problem (a bug??) in deserializing a layer with a picture element, encountering the following error ".   

Please advise.  Thank you in advance.

ERROR:
com.ait.lienzo.client.core.shape.json.validators.ValidationException: .fromString.children[1].attributes.url - value should be a URL
	at com.ait.lienzo.client.core.shape.json.JSONDeserializer.fromString(JSONDeserializer.java:77)
"

The same problem occurs when I tried the URL serialization mode as well, in another webapp where I can get the picture loaded from an external server.
_________________________________________
Here is the sample code to reproduce the problem:
package com.xxxxx.consumer.examples.client;

import java.util.logging.Logger;

import com.ait.lienzo.client.core.image.PictureLoadedHandler;
import com.ait.lienzo.client.core.shape.Layer;
import com.ait.lienzo.client.core.shape.Picture;
import com.ait.lienzo.client.core.shape.Text;
import com.ait.lienzo.client.core.shape.json.JSONDeserializer;
import com.ait.lienzo.client.widget.LienzoPanel;
import com.ait.lienzo.shared.core.types.Color;
import com.ait.lienzo.shared.core.types.ImageSerializationMode;
import com.google.gwt.core.client.EntryPoint;
import com.google.gwt.core.client.GWT;
import com.google.gwt.user.client.ui.RootPanel;

public class MyPicture implements EntryPoint {	
	private LienzoPanel panel;
	private Layer layer;
	private static String NEWLINE = "\n";	
	String url = GWT.getModuleBaseURL() + "images/raindrop.png";	
	Logger logger = Logger.getLogger("MyPicture");
	
	@Override
	public void onModuleLoad() {
		// PRELIMINARIES
		panel = new LienzoPanel(500, 600);
		layer = new Layer();
		panel.add(layer);
		RootPanel.get().add(panel);

		Text text = createText("Hello Lienzo", 100, 200);
		layer.add(text);
		layer.draw();
		logger.info("text drawn");
		
		new Picture(url, true, null).onLoaded(new PictureLoadedHandler() {         
            @Override  
            public void onPictureLoaded(Picture picture) {
            	picture.setDraggable(true);
            	picture.setImageSerializationMode(ImageSerializationMode.DATA_URL);
                layer.add(picture);  
                layer.draw(); 
        		logger.info("picture drawn");
        		testLayer(layer);
            }  
        });								
	}
	
	private Text createText(String textString, double xPos, double yPos) {
		// add a text shape at x-y coordinates
        Text text = new Text(textString, "Verdana, sans-serif", "italic bold", 40);
        text.setX(xPos).setY(yPos);
        text.setFillColor(Color.getRandomHexColor());
        text.setDraggable(true);
        return text;
	}
	
	private void testLayer(Layer layer){
		String json = layer.toJSONString();
		logger.info(">>>>>>>>>> LAYER serialized into json: " + NEWLINE + json + NEWLINE);
	 	Layer layerCopy = null;
		try {
			layerCopy = (Layer) JSONDeserializer.getInstance().fromString(json);
		} catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		logger.info("<<<<<<<<<< LAYER deserialized back to: " + NEWLINE + layerCopy + NEWLINE);
	}	
}

-------------------------------
Here are the outputs from Console:

Mon Feb 23 14:55:58 EST 2015 MyPicture
INFO: text drawn
Mon Feb 23 14:55:59 EST 2015 MyPicture
INFO: picture drawn
Mon Feb 23 14:55:59 EST 2015 MyPicture
INFO: >>>>>>>>>> LAYER serialized into json: 
{"type":"Layer", "attributes":{}, "children":[{"type":"Text", "attributes":{"text":"Hello Lienzo", "fontFamily":"Verdana, sans-serif", "fontStyle":"italic bold", "fontSize":40, "x":100, "y":200, "fill":"#48EB9E", "draggable":true}},{"type":"Picture", "attributes":{"url":"data:image/png;base64,iVBORw0KGgoAAA.......", "listening":true, "draggable":true, "serializationMode":"data-url"}}]}

com.ait.lienzo.client.core.shape.json.validators.ValidationException: .fromString.children[1].attributes.url - value should be a URL
	at com.ait.lienzo.client.core.shape.json.JSONDeserializer.fromString(JSONDeserializer.java:77)
	at com.helpswitch.consumer.examples.client.MyPicture.testLayer(MyPicture.java:82)
	at com.helpswitch.consumer.examples.client.MyPicture.access$1(MyPicture.java:77)
	at com.helpswitch.consumer.examples.client.MyPicture$1.onPictureLoaded(MyPicture.java:55)
	at com.ait.lienzo.client.core.shape.Picture$1.onImageShapeLoaded(Picture.java:1595)
	at com.ait.lienzo.client.core.shape.Picture$1.onImageShapeLoaded(Picture.java:1)
	at com.ait.lienzo.client.core.image.ImageProxy.doneLoading(ImageProxy.java:615)
	at com.ait.lienzo.client.core.image.ImageProxy.doInitialize(ImageProxy.java:403)
	at com.ait.lienzo.client.core.image.ImageProxy.access$8(ImageProxy.java:351)
	at com.ait.lienzo.client.core.image.ImageProxy$1.onLoad(ImageProxy.java:308)
	at com.ait.lienzo.client.core.image.ImageLoader$1.onLoad(ImageLoader.java:61)
	at com.google.gwt.event.dom.client.LoadEvent.dispatch(LoadEvent.java:56)
	at com.google.gwt.event.dom.client.LoadEvent.dispatch(LoadEvent.java:1)
	at com.google.gwt.event.shared.GwtEvent.dispatch(GwtEvent.java:1)
	at com.google.web.bindery.event.shared.EventBus.dispatchEvent(EventBus.java:40)
	at com.google.web.bindery.event.shared.SimpleEventBus.doFire(SimpleEventBus.java:193)
	at com.google.web.bindery.event.shared.SimpleEventBus.fireEvent(SimpleEventBus.java:88)
	at com.google.gwt.event.shared.HandlerManager.fireEvent(HandlerManager.java:127)
	at com.google.gwt.user.client.ui.Widget.fireEvent(Widget.java:129)
	at com.google.gwt.event.dom.client.DomEvent.fireNativeEvent(DomEvent.java:125)
	at com.google.gwt.user.client.ui.Widget.onBrowserEvent(Widget.java:177)
	at com.google.gwt.user.client.ui.Image.onBrowserEvent(Image.java:816)
	at com.google.gwt.user.client.DOM.dispatchEventImpl(DOM.java:1467)
	at com.google.gwt.user.client.DOM.dispatchEvent(DOM.java:1406)
	at com.google.gwt.user.client.impl.DOMImplStandard.dispatchEvent(DOMImplStandard.java:323)
	at com.google.gwt.user.client.impl.DOMImplStandard.dispatchUnhandledEvent(DOMImplStandard.java:343)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.google.gwt.dev.shell.MethodAdaptor.invoke(MethodAdaptor.java:103)
	at com.google.gwt.dev.shell.MethodDispatch.invoke(MethodDispatch.java:71)
	at com.google.gwt.dev.shell.OophmSessionHandler.invoke(OophmSessionHandler.java:172)
	at com.google.gwt.dev.shell.BrowserChannelServer.reactToMessagesWhileWaitingForReturn(BrowserChannelServer.java:338)
	at com.google.gwt.dev.shell.BrowserChannelServer.invokeJavascript(BrowserChannelServer.java:219)
	at com.google.gwt.dev.shell.ModuleSpaceOOPHM.doInvoke(ModuleSpaceOOPHM.java:136)
	at com.google.gwt.dev.shell.ModuleSpace.invokeNative(ModuleSpace.java:576)
	at com.google.gwt.dev.shell.ModuleSpace.invokeNativeObject(ModuleSpace.java:284)
	at com.google.gwt.dev.shell.JavaScriptHost.invokeNativeObject(JavaScriptHost.java:91)
	at com.google.gwt.core.client.impl.Impl.apply(Impl.java)
	at com.google.gwt.core.client.impl.Impl.entry0(Impl.java:347)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.google.gwt.dev.shell.MethodAdaptor.invoke(MethodAdaptor.java:103)
	at com.google.gwt.dev.shell.MethodDispatch.invoke(MethodDispatch.java:71)
	at com.google.gwt.dev.shell.OophmSessionHandler.invoke(OophmSessionHandler.java:172)
	at com.google.gwt.dev.shell.BrowserChannelServer.reactToMessages(BrowserChannelServer.java:293)
	at com.google.gwt.dev.shell.BrowserChannelServer.processConnection(BrowserChannelServer.java:547)
	at com.google.gwt.dev.shell.BrowserChannelServer.run(BrowserChannelServer.java:364)
	at java.lang.Thread.run(Thread.java:744)
Mon Feb 23 14:55:59 EST 2015 MyPicture
INFO: <<<<<<<<<< LAYER deserialized back to: 
null

**Comment by SprocketNYC** [comment]
A small issue with the data URL validity checker, I should be able to fix it tomorrow

**Comment by yskwong** [comment]
Great.  Thank you very much.

Please note that the problem occurs in both serialization modes: data URL, or just URL.

**Comment by SprocketNYC** [comment]
what version of GWT are you using?

**Comment by yskwong** [comment]
I am on GWT 2.6.0, with Lienzo 2.0.55 snapshot.  Thanks.

**Comment by yskwong** [comment]
I upgraded Lienzo to 2.0.63, the latest snapshot.  The app is still not working, getting the same error in deserializing picture.

Could it be really be due to GWT 2.6.0?? I hesitate to upgrade GWT in the midst of a project.

**Comment by yskwong** [comment]
I upgraded my Eclipse to use gwt 2.7.0, and run the same app.  The same problem/error exists, independent of whether I am using gwt 2.6.0 or 2.7.0.  It is curious that the sample in the KitchenSInk works, but the source there does not include the codes for deserialization,  So I just wonder if you have a chance to reproduce the problem reported here.  Thank you.

**Comment by SprocketNYC** [comment]
It's an issue between the JSON Parser and GWT's JSON package, big data URL's seem to be, not just blank, missing...

back to it tmrw

**Comment by SprocketNYC** [comment]
try 2.0.71.SNAPSHOT

**Comment by yskwong** [comment]
I tried out Lienzo 2.0.71.  Deserialization of pictures works in Chrome (40.0.2214.115), but not working in either FireFox (version 26) or Safari (version 8.0.3).  Tests were done using DATA URL for serialization/deseralization.

I tried testing using URL on another web app,  but encountered an incompatibility issue in Lienzo 2.0.71 with older snapshots (such as 2.0.55 or 2.0.63): the static method "FactoryRegistry.getInstance()" is no longer available in 2.0.71.  Is that an intentional API change or an oversight??  Please advise.

**Comment by SprocketNYC** [comment]
well, on Safari and Firefox I get similar errors, the Data URL is REALLY BIG and the JavaScript parser refuses to deal with it, this is the 512x512 image Lena, size in chars is 1,130,163 !!!!

"LENGTH=1130163" lienzoks-0.js:12278:4
"UNCAUGHT EXCEPTION " lienzoks-0.js:97945:2
"com.google.gwt.core.client.JavaScriptException: (IndexSizeError) : Index or size is negative or greater than the allowed amount

Dunno what I can do about that.... thinking... compress it? break it up into an array of strings ??? Maybe something else wrong, that is WAY bigger than Chrome. It might have to do with the backing store pixel ratiio.

It could be the WHOLE JSON String is too big to parse.

Anyway, FactoryRegistry.getInstance() was changed to .get(), just to align it with the same pattern in the rest of the API. The Shape Factories as defined in LienzoCorePlugin.java were refactored quite extensively, they are now NOT all allocated up front, they are lazy created only when they are needed by the JSON Deserializer.

**Comment by yskwong** [comment]
I will upgrade to Lienzo 2.0.71 to test out the usage of URL in serialization, and avoid data URL, see if that would avoid the problem.

What is the equivalent of registerFactory method now in the API re FactoryRegistry?  Thanks.

**Comment by SprocketNYC** [comment]
The FactoryRegistry is now gone, you must use the Lienzo plug-in system to add factories, it's rather easy, create a plugin:

```java
import com.ait.lienzo.client.core.config.AbstractLienzoCorePlugin;
import com.ait.lienzo.client.core.shape.json.IFactory;
import com.ait.lienzo.shared.java.util.function.Supplier;

public final class LienzoChartsPlugin extends AbstractLienzoCorePlugin
{
    public LienzoChartsPlugin()
    {
        addFactorySupplier(ChartNodeType.PIE_CHART, new Supplier<IFactory<?>>()
        {
            @Override
            public IFactory<?> get()
            {
                return new PieChart.PieChartFactory();
            }
        });
    }

    @Override
    public final String getNameSpace()
    {
        return "LienzoCharts";
    }

    @Override
    public final String getVersion()
    {
        return "1.0";
    }
}
```

Then register the plugin:

```java
LienzoCore.get().addPlugin(new LienzoChartsPlugin());
```

This model is VITALLY important for future plans. Also the getInstance() is changed to get() in JSONDeserializer

Lienzo is still SNAPSHOTS, I try not to make breaking changes, but ... sometimes I do, but they are minor, in places in code people usually either don't use much or SHOULDN'T be accessing in a way that it was not meant to be.

Anyway 2.0.72 is pushed, it's... sometimes better on Picture serialization... I can't get Firefox to stop producing such HUGE data URL's

**Comment by yskwong** [comment]
Thank you very much for the assistance.

I will try out both the Lienzo plug-in for factories, and the picture deserialization, and give you the feedback afterward.

Best regards,

**Comment by SprocketNYC** [comment]
It's still really bad on FF, I tried changing the DataURLType to jpeg and it still only send back png!!!!

**Comment by yskwong** [comment]
So long as deserialization of pictures work with URL, I can live with it and avoid using DATA URL. Can you confirm if the latest snapshots (2.0.71, 72) works with URL?  I could not check right now because of the incompatibility issue for factory registry that my app is using.

I am not too sure if size of the bitmap is really the problem.  I tried a picture 128x128 (that serialized into png with about under 8K of characters.  Chrome works fine, but not FF or Safari (using 2.0.71).

Thanks again.

**Comment by SprocketNYC** [comment]
Regular URL's are fine, AFAIK, been testing them. There was an error in the URLValidator class.

**Comment by yskwong** [comment]
Thank you.

**Comment by SprocketNYC** [comment]
So, regular URL's work fine, data url's have some issues, looks like Chrome accepts 2M, Safari less but not great, and Firefox, 800kb

Can't do much else with it   :-(  Firefox just sucks.

**Comment by yskwong** [comment]
Two feedbacks:

1)	I switched everything to use URL, and deserialization of pictures in URL mode works fine in 2.0.71 or 72.

2)	Size of the picture might be an issue to explain the difference in behavior of Chome vs FF or Safari.  But I really doubt that is the root cause.  

When I reported this deserializaton problem, the picture I tried to serialize and deserialize is about 8KB.  It is hardly big.

To convince myself size is not necessary the issue, I tried the same code used in reporting the issue #56, but with a png icon (copy.png icon) which is only 1KB.

Still the same problem in picture deserialization:  works fine in Chrome (using 2.0.72) but NOT in FF or Safari.

This is not stopping my work, but might point at a hidden bug(s) somewhere:)  Thanks.
